### PR TITLE
Contributing Guide

### PR DESCRIPTION
**Pull request by VishrutiBuddhadev**

### CODE REVIEW

File: docs/CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
+# Contributing to infoblox.universal_ddi collection
+
+Welcome to the Universal DDI Ansible Repository! We are excited that you are interested in contributing to our project. This document outlines the process for contributing to the repository and how you can make submissions that add value and are in harmony with the current structure.
+
+## Workflow Summary
+
+1. [Clone the repository](#cloning)
+2. [Make the desired Code Change](#writing-new-code)
```
Comment (unasra) [inline docs/CONTRIBUTING.md]:
```suggestion
2. [Implementing the Code Changes](#writing-new-code)
```

File: docs/CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
+# Contributing to infoblox.universal_ddi collection
+
+Welcome to the Universal DDI Ansible Repository! We are excited that you are interested in contributing to our project. This document outlines the process for contributing to the repository and how you can make submissions that add value and are in harmony with the current structure.
```
Comment (unasra) [inline docs/CONTRIBUTING.md]:
```suggestion
Welcome to the Universal DDI Ansible Repository ! We are excited that you are interested in contributing to our project. This document outlines the process for contributing to the repository and how you can make submissions that add value and are in harmony with the current structure and coding standards.
```

File: docs/CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+## Workflow Summary
+
+1. [Clone the repository](#cloning)
```
Comment (unasra) [inline docs/CONTRIBUTING.md]:
```suggestion
1. [Setup](#cloning)
```

File: docs/CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+2. [Make the desired Code Change](#writing-new-code)
+3. [Run integration tests locally and ensure that they pass](#writing-integration-tests)
```
Comment (unasra) [inline docs/CONTRIBUTING.md]:
```suggestion
3. [Testing the Code Changes](#writing-integration-tests)
```

File: docs/CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+4. [Create a PR](#pr-creation)
+
+## Cloning
```
Comment (unasra) [inline docs/CONTRIBUTING.md]:
```suggestion
## Setup
```

File: docs/CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+The `ansible-test` command expects that the repository is in a directory that matches it's collection,
+under a directory `ansible_collections`. Clone ensuring that hierarchy:
```
Comment (unasra) [inline docs/CONTRIBUTING.md]:
```suggestion
under the directory `ansible_collections`. Clone the repository from Github ensuring the hierarchy:
```

File: docs/CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+```shell
+mkdir -p $TARGET_DIR/ansible_collections/infoblox
+git clone <url> $TARGET_DIR/ansible_collections/infoblox/universal_ddi
+```
+
+The python requirements must be installed separately. It is recommended to use a venv to keep the dependencies local.
```
Comment (unasra) [inline docs/CONTRIBUTING.md]:
```suggestion
The python requirements must be installed separately. It is recommended to use a Virtual Environment to isolate dependencies and maintain a clean setup.
```

File: docs/CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+The python requirements must be installed separately. It is recommended to use a venv to keep the dependencies local. 
+Set up your Python virtual environment
+
+```shell
+cd $TARGET_DIR/ansible_collections/infoblox/universal_ddi
+python3 -m venv .venv
+source .venv/bin/activate
+pip3 install ansible
+pip3 install -r requirements.txt
```
Comment (unasra) [inline docs/CONTRIBUTING.md:71]: Need to add one more command 
` pip install -r test-requirements.txt `

File: docs/CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+```
+
+You can run ```ansible-test sanity``` to test if your collection is ready for development
+
+## Writing New Code
```
Comment (unasra) [inline docs/CONTRIBUTING.md]:
```suggestion
## Implementing the code changes
```

File: docs/CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+### Writing New Modules
```
Comment (unasra) [inline docs/CONTRIBUTING.md]:
```suggestion
### Developing New Modules
```

File: docs/CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+ 
+Create two files for each module:
+- `module_name.py`: The main module code.
```
Comment (unasra) [inline docs/CONTRIBUTING.md]:
```suggestion
- `module_name.py`: Module for performing CRUD operations
```

File: docs/CONTRIBUTING.md [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+- `module_name_info.py`: For gathering module resource information.
```
Comment (unasra) [inline docs/CONTRIBUTING.md]:
```suggestion
- `module_name_info.py`: Module to List / Read the object
```

File: docs/CONTRIBUTING.md [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+
+Place these files under `/plugins/modules`.
+
+#### Documentation Examples and Return section inside modules
+
+**Documentation Section** Each module file should start with a comprehensive documentation section. [Module format and documentation](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#module-format-and-documentation) guide gives  details about how to document.
+
+**Examples Section** Provides practical usage examples demonstrating standard and edge case scenarios.
+
+**Return Section** Clearly document what values the module returns.
```
Comment (unasra) [inline docs/CONTRIBUTING.md]:
```suggestion
**Return Section** -  It defines what data is supposed to be returned and how it is structured.
```

File: docs/CONTRIBUTING.md [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+
+
+#### Expected test criteria:
+- Resource creation under check mode
+- Resource creation
+- Resource creation again (idempotency)
+- Resource deletion under check mode
+- Resource deletion
+- Resource deletion again (idempotency)
+- Resource creation with additional parameters
```
Comment (unasra) [inline docs/CONTRIBUTING.md]:
```suggestion
- Resource updation with all additional attributes under the object
```

File: docs/CONTRIBUTING.md [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+- Resource modification
```
Comment (unasra) [inline docs/CONTRIBUTING.md]: Not Required as such , the line above would cover this

File: docs/CONTRIBUTING.md [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+
+#### Running Tests Locally
```
Comment (unasra) [inline docs/CONTRIBUTING.md]:
```suggestion
#### Testing the code changes
```

File: docs/CONTRIBUTING.md [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+
+#### Documentation Examples and Return section inside modules
+
+**Documentation Section** Each module file should start with a comprehensive documentation section. [Module format and documentation](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#module-format-and-documentation) guide gives  details about how to document.
+
+**Examples Section** Provides practical usage examples demonstrating standard and edge case scenarios.
+
+**Return Section** Clearly document what values the module returns.
+
+### Writing Integration Tests
```
Comment (unasra) [inline docs/CONTRIBUTING.md]: Can you shift this section to testing the code ? 
This way we can separate development and testing.

File: docs/CONTRIBUTING.md [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+- Resource creation again (idempotency)
+- Resource deletion under check mode
+- Resource deletion
+- Resource deletion again (idempotency)
+- Resource creation with additional parameters
+- Resource modification
+
+#### Running Tests Locally
+
+For local testing, ensure you set up your integration test configuration
```
Comment (unasra) [inline docs/CONTRIBUTING.md]: Can you place this under a sub section of `Testing the Code Changes ` , naming it either `Configuration` .

File: docs/CONTRIBUTING.md [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+1. Navigate to `tests/integration` folder
+2. Create the integration_config.yaml file with necessary environment-specific configurations such as portal_url and portal_key as shown below.
+
+   ```yaml
```
Comment (unasra) [inline docs/CONTRIBUTING.md]: Add the module_defaults part we have in the integration tests and also add a line stating that we define the configuration details under universal_ddi/all so the config values get added to all the calls in the playbook 
(please rephrase it )

File: docs/CONTRIBUTING.md [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+- The PR will trigger some check actions in Github. The PR can be only merged when these tests pass. You can also test your PR locally with these commands. 
+
+```shell
+ansible-test sanity
+```
+
+```shell
+ansible-test integration
+```
+
```
Comment (unasra) [inline docs/CONTRIBUTING.md:208]: One more line about how the PR would go under review by maintainers of this repository and how the contributor is expected to make changes as per the review for the PR to get merged.

File: docs/CONTRIBUTING.md (hunk 9 above)
Comment (unasra) [inline docs/CONTRIBUTING.md]: Can you add a section or a sub section before this , giving the link to the Bloxone API documentation stating that the user can refer to this link for all information and attributes related to the module they want to contribute ?

File: docs/CONTRIBUTING.md [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+#### Expected test criteria:
+- Resource creation under check mode
+- Resource creation
+- Resource creation again (idempotency)
+- Resource deletion under check mode
+- Resource deletion
+- Resource deletion again (idempotency)
+- Resource creation with additional parameters
+- Resource modification
+
```
Comment (unasra) [inline docs/CONTRIBUTING.md:138]: Also before running the tests , a sub section is required for formatting , basically saying that the contributor needs to run the `make lint` command to ensure proper formatting
We can also explain that these commands run , flynt ,black and isort python libraries and those are required to be installed . These have already been added in test-requirements.txt

File: docs/CONTRIBUTING.md [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+## Workflow Summary
+
+1. [Clone the repository](#cloning)
+2. [Make the desired Code Change](#writing-new-code)
+3. [Run integration tests locally and ensure that they pass](#writing-integration-tests)
+4. [Create a PR](#pr-creation)
+
+## Cloning
+
+The `ansible-test` command expects that the repository is in a directory that matches it's collection,
```
Comment (unasra) [inline docs/CONTRIBUTING.md]: Also , we can explain this structure like a proper directory diagram 

image

U can refer here for futher examples , https://github.com/IBM/z_ansible_collections_samples/blob/main/docs/share/contribution-guidelines.md?plain=1

File: docs/CONTRIBUTING.md [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+
+### Developing New Modules
+ 
+Create two files for each module:
+- `module_name.py`: Module for performing CRUD operations.
+- `module_name_info.py`: Module to List / Read the object.
+
+Place these files under `/plugins/modules`.
+
+#### Documentation Examples and Return section inside modules
```
Comment (unasra) [inline docs/CONTRIBUTING.md]:
```suggestion
#### Documentation, Examples and Return section inside modules
```

File: docs/CONTRIBUTING.md [hunk 24]
```diff
@@ -0,0 +1,145 @@
...
+
+- Start by adding tests in tests/integration/targets/MODULE_NAME for your new feature.
+
+- Each module must also be added to the all group in meta/runtime.yml. This will allow to reuse common variables like portal_key in the test. 
+
+### Expected test criteria
+
+- Resource creation under check mode
+- Resource creation
+- Resource creation again (idempotency)
```
Comment (unasra) [inline docs/CONTRIBUTING.md]: Removing the delete wasnt necessary , we still need to mention that 
Delete Check mode , delete and delete idempotent are required.

Also adda a line that for a more detailed example , the contributor can one at one of the examples in the tests section.

File: docs/CONTRIBUTING.md [hunk 25]
```diff
@@ -0,0 +1,145 @@
...
+
+## Testing the Code Changes
+
+### Writing Integration Tests
+
+- All modules MUST have integration tests for new features. Bug fixes for modules that currently have integration tests SHOULD have tests added.
+
+- Start by adding tests in tests/integration/targets/MODULE_NAME for your new feature.
+
+- Each module must also be added to the all group in meta/runtime.yml. This will allow to reuse common variables like portal_key in the test.
```
Comment (unasra) [inline docs/CONTRIBUTING.md]: Make this the first point of this section

File: docs/CONTRIBUTING.md [hunk 26]
```diff
@@ -0,0 +1,145 @@
...
+
+**Return Section** It defines what data is supposed to be returned and how it is structured.
+
+## Testing the Code Changes
+
+### Writing Integration Tests
+
+- All modules MUST have integration tests for new features. Bug fixes for modules that currently have integration tests SHOULD have tests added.
+
+- Start by adding tests in tests/integration/targets/MODULE_NAME for your new feature.
```
Comment (unasra) [inline docs/CONTRIBUTING.md]:
```suggestion
- Start by adding tests in tests/integration/targets/MODULE_NAME for your new feature. The tests are supposed to be added for all writable attributes of the object. The complete list of attributes can be found in the API documentation.
```

File: docs/CONTRIBUTING.md (hunk 22 above)
Comment (unasra) [reply docs/CONTRIBUTING.md]: Also the coding Style guidelines provided in the link can be added.

File: docs/CONTRIBUTING.md (hunk 21 above)
Comment (unasra) [reply docs/CONTRIBUTING.md:138]: https://github.com/IBM/z_ansible_collections_samples/blob/main/docs/share/contribution-guidelines.md#linting

File: docs/CONTRIBUTING.md (hunk 24 above)
Comment (unasra) [reply docs/CONTRIBUTING.md]: To make it better , this can be referred 
https://github.com/openstack/ansible-collections-openstack/blob/master/docs/contributing.md#examples

**Comment by unasra** [comment]
Maybe we can also add a license for contributing , this could be a good reference. 

https://github.com/aruba/aoscx-ansible-collection/blob/master/CONTRIBUTING.md#licensing

File: docs/CONTRIBUTING.md [hunk 27]
```diff
@@ -0,0 +1,218 @@
+# Contributing to infoblox.universal_ddi collection
+
+Welcome to the Universal DDI Ansible Repository ! We are excited that you are interested in contributing to our project. This document outlines the process for contributing to the repository and how you can make submissions that add value and are in harmony with the current structure and coding standards.
```
Comment (unasra) [inline docs/CONTRIBUTING.md]:
```suggestion
Welcome to the Universal DDI Ansible Repository ! We're thrilled about your interest in contributing to our project. This document outlines the process for contributing to the repository and how you can make submissions that add value and are in harmony with the current structure and coding standards.
```

File: docs/CONTRIBUTING.md [hunk 28]
```diff
@@ -0,0 +1,218 @@
...
+        name: "{{ name }}"
+        state: "absent"
+      ignore_errors: true
+    ```
+### Expected test criteria
+
+- Resource creation under check mode
+- Resource creation
+- Resource creation again (idempotency)
+- Resource updation with all additional attributes under the object
```
Comment (unasra) [inline docs/CONTRIBUTING.md]: Resource updation should come after the line "Resource deletion again (idempotency)"

File: docs/CONTRIBUTING.md [hunk 29]
```diff
@@ -0,0 +1,218 @@
...
+- Resource deletion under check mode
+- Resource deletion
+- Resource deletion again (idempotency)
+
+For more detailed example, refer to [/tests/integration/targets/ipam_ip_space/tasks/main.yml](../tests/integration/targets/ipam_ip_space/tasks/main.yml)
+
+### Configuration
```
Comment (unasra) [inline docs/CONTRIBUTING.md]: Configuration Section should come before Writing Integration Tests , under Testing Code Changes

File: docs/CONTRIBUTING.md [hunk 30]
```diff
@@ -0,0 +1,218 @@
...
+│   ├── module_utils/
+│   ├── modules/
+│   └── plugin_utils/
+├── tests
+│   ├── integration/
+│   │   └── targets/
+│   └── unit/
+└── venv/
+```
+
```
Comment (unasra) [inline docs/CONTRIBUTING.md:48]: We can mention here that the user should create a fork and the changes should be made to their fork.

File: docs/CONTRIBUTING.md [hunk 31]
```diff
@@ -0,0 +1,218 @@
...
+Clone the repository from GitHub ensuring the hierarchy.
+```shell
+mkdir -p $TARGET_DIR/ansible_collections/infoblox
+git clone <url> $TARGET_DIR/ansible_collections/infoblox/universal_ddi
```
Comment (unasra) [inline docs/CONTRIBUTING.md]: Once the fork line is added , the `<url>` can replaced with the actual URL and we can add a placeholder for user name.

File: docs/CONTRIBUTING.md [hunk 32]
```diff
@@ -0,0 +1,218 @@
...
+          name: "{{ name }}"
+          state: "present"
+          .....
+   ``` 
+
+- Use well-defined and easily readable variable names.
+
+     ```name: "test-ip-space"```
+
+- If your playbook has many debug statements consider using the `verbosity` parameter that allows you to control when certain messages appear.
```
Comment (unasra) [inline docs/CONTRIBUTING.md]: I feel that this guideline is not required . Under the "Running Tests" , this has already been mentioned and using -vv is better because this way contributor might leave debug statements in code.

File: docs/CONTRIBUTING.md [hunk 33]
```diff
@@ -0,0 +1,218 @@
...
+    ```yaml
+  always:
+    # Cleanup if the test fails
+    - name: "Delete IP Space"
+      infoblox.universal_ddi.ipam_ip_space:
+        name: "{{ name }}"
+        state: "absent"
+      ignore_errors: true
+    ```
+### Expected test criteria
```
Comment (unasra) [inline docs/CONTRIBUTING.md]: This section should be moved before coding style guidelines.

File: docs/CONTRIBUTING.md [hunk 34]
```diff
@@ -0,0 +1,211 @@
...
+
+- Resource creation under check mode
+- Resource creation
+- Resource creation again (idempotency)
+- Resource deletion under check mode
+- Resource deletion
+- Resource deletion again (idempotency)
+- Resource updation with all additional attributes under the object
+
+For more detailed example, refer to [/tests/integration/targets/ipam_ip_space/tasks/main.yml](../tests/integration/targets/ipam_ip_space/tasks/main.yml)
```
Comment (unasra) [inline docs/CONTRIBUTING.md]:
```suggestion
For more detailed example, refer to any of test files(../tests/integration/targets/ipam_ip_space/tasks/main.yml)
```
We can have the link on the files keyword here.

File: docs/CONTRIBUTING.md [hunk 35]
```diff
@@ -0,0 +1,211 @@
...
+Welcome to the Universal DDI Ansible Repository ! We're thrilled about your interest in contributing to our project. This document outlines the process for contributing to the repository and how you can make submissions that add value and are in harmony with the current structure and coding standards.
+## Workflow Summary
+
+1. [Setup](#setup)
+2. [Implementing the Code Changes](#implementing-the-code-changes)
+3. [Testing the Code Changes](#testing-the-code-changes)
+4. [Creating a PR](#creating-a-pr)
+
+## Setup
+
```
Comment (unasra) [inline docs/CONTRIBUTING.md:25]: Lets divide this setup in two segments 
Directory Structure and Installing requirements

File: docs/CONTRIBUTING.md [hunk 36]
```diff
@@ -0,0 +1,211 @@
...
+└── venv/
+```
+Before making any changes, it's essential to fork the repository. This allows you to work on your own copy of the project and makes it easy to contribute your changes back to the main repository through a pull request.
+
+Clone the repository from GitHub ensuring the hierarchy.
+```shell
+mkdir -p $TARGET_DIR/ansible_collections/infoblox
+git clone https://github.com/<your_username>/universal-ddi-ansible.git $TARGET_DIR/ansible_collections/infoblox/universal_ddi
+```
+
```
Comment (unasra) [inline docs/CONTRIBUTING.md:54]: The Installing requirements can start from here . 
Also we should separate the setting up of venv and installation code into different code blocks , as some users might already have their venv setup .

File: docs/CONTRIBUTING.md [hunk 37]
```diff
@@ -0,0 +1,211 @@
...
+    - name: "Delete IP Space"
+      infoblox.universal_ddi.ipam_ip_space:
+        name: "{{ name }}"
+        state: "absent"
+      ignore_errors: true
+    ```
+
+### Linting
+
+To facilitate consistent coding style across the project, please run the following command before running the tests
```
Comment (unasra) [inline docs/CONTRIBUTING.md]:
```suggestion
To facilitate consistent formatting across the project, please run the following command before running the tests
```

File: docs/CONTRIBUTING.md [hunk 38]
```diff
@@ -0,0 +1,231 @@
...
+ 
+Create two files for each module:
+- `module_name.py`: Module for performing CRUD operations.
+- `module_name_info.py`: Module to List / Read the object.
+
+Place these files under `/plugins/modules`.
+
+#### Documentation, Examples and Return section inside modules
+
+**Documentation Section** Each module file should start with a comprehensive documentation section. [Module format and documentation](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#module-format-and-documentation) guide gives  details about how to document.
```
Comment (unasra) [inline docs/CONTRIBUTING.md]:
```suggestion
**Documentation Section** : Each module file should start with a comprehensive documentation section. [Module format and documentation](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#module-format-and-documentation) guide gives  details about how to document.
```

Applies for other two points too.

File: docs/CONTRIBUTING.md [hunk 39]
```diff
@@ -0,0 +1,231 @@
...
+│   ├── integration/
+│   │   └── targets/
+│   └── unit/
+└── venv/
+```
+Before making any changes, it's essential to fork the repository. This allows you to work on your own copy of the project and makes it easy to contribute your changes back to the main repository through a pull request.
+
+Clone the repository from GitHub ensuring the hierarchy.
+```shell
+mkdir -p $TARGET_DIR/ansible_collections/infoblox
```
Comment (unasra) [inline docs/CONTRIBUTING.md]:
```suggestion
mkdir -p $TARGET_DIR/ansible_collections/infoblox/universal_ddi
```

File: docs/CONTRIBUTING.md [hunk 40]
```diff
@@ -0,0 +1,231 @@
+# Contributing to infoblox.universal_ddi collection
+
+Welcome to the Universal DDI Ansible Repository ! We're thrilled about your interest in contributing to our project. This document outlines the process for contributing to the repository and how you can make submissions that add value and are in harmony with the current structure and coding standards.
+
+## Workflow Summary
```
Comment (unasra) [inline docs/CONTRIBUTING.md]: This loogs good but , Better Make this as index only and all the headings , Workflow Summary does not make sense anymore sense we added all the links

**Review by unasra — APPROVED** [review]
